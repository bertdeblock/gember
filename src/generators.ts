import { generateDocument } from "./generate-document.js";
import { DocumentName } from "./types.js";

export function generateComponent(
  name: string,
  cwd: string,
  options: { gts: boolean } = { gts: false },
) {
  return generateDocument(DocumentName.Component, name, cwd, options);
}

export function generateHelper(
  name: string,
  cwd: string,
  options: { ts: boolean } = { ts: false },
) {
  return generateDocument(DocumentName.Helper, name, cwd, options);
}

export function generateModifier(
  name: string,
  cwd: string,
  options: { ts: boolean } = { ts: false },
) {
  return generateDocument(DocumentName.Modifier, name, cwd, options);
}

export function generateService(
  name: string,
  cwd: string,
  options: { ts: boolean } = { ts: false },
) {
  return generateDocument(DocumentName.Service, name, cwd, options);
}
