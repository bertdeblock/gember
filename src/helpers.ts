import type { EmberPackageJson } from "./types.js";

export function isV1Addon(packageJson: EmberPackageJson): boolean {
  if (isAddon(packageJson)) {
    const { version } = packageJson["ember-addon"] ?? {};

    return version === 1 || version === undefined;
  }

  return false;
}

export function isV2Addon(packageJson: EmberPackageJson): boolean {
  if (isAddon(packageJson)) {
    const { version } = packageJson["ember-addon"] ?? {};

    return version === 2;
  }

  return false;
}

function isAddon(packageJson: EmberPackageJson): boolean {
  if (Array.isArray(packageJson.keywords)) {
    return packageJson.keywords.includes("ember-addon");
  }

  return false;
}
