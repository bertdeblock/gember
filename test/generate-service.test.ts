import { afterEach, it } from "vitest";
import { generateService } from "../src/generators.ts";
import { Package } from "./helpers.ts";

let pkg: Package;

afterEach(() => pkg.cleanUp());

it("generates a `.js` service", async (ctx) => {
  pkg = await Package.create("v2-addon");

  await generateService("foo", pkg.path);

  const content = await pkg.readFile("src/services/foo.js");

  ctx.expect(content).toMatchSnapshot();
});

it("generates a `.ts` service", async (ctx) => {
  pkg = await Package.create("v2-addon");

  await generateService("foo", pkg.path, { typescript: true });

  const content = await pkg.readFile("src/services/foo.ts");

  ctx.expect(content).toMatchSnapshot();
});

it("generates a `.js` service at a custom path", async (ctx) => {
  pkg = await Package.create("v2-addon");

  await generateService("foo", pkg.path, { path: "src/-private" });

  const content = await pkg.readFile("src/-private/foo.js");

  ctx.expect(content).toMatchSnapshot();
});

it("generates a `.ts` service at a custom path", async (ctx) => {
  pkg = await Package.create("v2-addon");

  await generateService("foo", pkg.path, {
    path: "src/-private",
    typescript: true,
  });

  const content = await pkg.readFile("src/-private/foo.ts");

  ctx.expect(content).toMatchSnapshot();
});

it("generates a nested `.js` service", async (ctx) => {
  pkg = await Package.create("v2-addon");

  await generateService("foo/bar", pkg.path);

  const content = await pkg.readFile("src/services/foo/bar.js");

  ctx.expect(content).toMatchSnapshot();
});
