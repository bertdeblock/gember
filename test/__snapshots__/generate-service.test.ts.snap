// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generates a \`.js\` service 1`] = `
"import Service from "@ember/service";

export default class FooService extends Service {}
"
`;

exports[`generates a \`.ts\` service 1`] = `
"import Service from "@ember/service";

export default class FooService extends Service {}

declare module "@ember/service" {
  interface Registry {
    "foo": FooService;
  }
}
"
`;
